Variables: Adds named capture groups to variable regex

yenalsnavaj opened the pull request:
This PR allows 'text' and 'value' named capture groups to be specified in the regular expression for variable queries so that distinct values and display text can be extracted. This is simliar the __text and __value functionality available if using a SQL datasource but should work for all datasources - especially Prometheus, which is the tested use case. 

Fixes #21076
Fixes #12039

yenalsnavaj committed:
Dashboard: Add named capture groups to variable query regex

Variable query regex are able to use 'text' and 'value' named capture
groups to allow for separate display text to be extracted from the
query result. e.g.

Using a regex of /foo="(?<text>[^"]+)|bar="(?<value>[^"]+)/g on a query
result of metric{foo="FOO", bar="BAR"} would result in the variable
value being set to 'BAR' but display text being set to 'FOO'

Resolves #21076

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=28625) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=28625) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=28625) it.</sub>

oddlittlebird commented on docs/sources/variables/filter-variables-with-regex.md:
```diff
@@ -79,4 +79,30 @@ demo.robustperception.io:9090
+
+## Filter and use named text and value capture groups
+
+This can be useful if the display text should differ from the value.
```
What can be useful? Please explain what "this"is that you are referring to.

oddlittlebird commented on docs/sources/variables/filter-variables-with-regex.md:
```diff
@@ -79,4 +79,30 @@ demo.robustperception.io:9090
+
+This can be useful if the display text should differ from the value.
+
+List of options:
```
I am confused. Are these options an example, or are they common to all users? This section needs more context, as do the Regex and Result below.

yenalsnavaj replied:
Sure. I was following the format of the examples earlier the file. Agree that it could be improved - documentation is not my strong point ;) I’ll rework with expanded examples that relate tot the use case in the issue that it fixes

oddlittlebird reviewed (commented):
I have a few suggestions for the documentation.

yenalsnavaj committed:
Improve regex capture group documentation

yenalsnavaj commented:
@oddlittlebird i've updated the documentation. Let me know what you think.

oddlittlebird commented on docs/sources/variables/filter-variables-with-regex.md:
````diff
@@ -79,4 +79,35 @@ demo.robustperception.io:9090
+/chip_name="(?<text>[^"]+)|chip="(?<value>[^"]+)/g
 ```
+
+Would produce the following dropdown:
````
```suggestion
Would produce the following drop-down list:
```

oddlittlebird commented on docs/sources/variables/filter-variables-with-regex.md:
```diff
@@ -79,4 +79,35 @@ demo.robustperception.io:9090
+
+## Filter and modify using named text and value capture groups
+
+Using named capture groups it is possible to capture separate 'text' and 'value' parts from the options returned by the variable query. This allows the variable dropdown to contain a friendly name for each value that can be selected.
```
```suggestion
Using named capture groups, you can capture separate 'text' and 'value' parts from the options returned by the variable query. This allows the variable drop-down list to contain a friendly name for each value that can be selected.
```

oddlittlebird commented on docs/sources/variables/filter-variables-with-regex.md:
```diff
@@ -79,4 +79,35 @@ demo.robustperception.io:9090
+
+Using named capture groups it is possible to capture separate 'text' and 'value' parts from the options returned by the variable query. This allows the variable dropdown to contain a friendly name for each value that can be selected.
+
+For example, when querying the `node_hwmon_chip_names` prometheus metric the `chip_name` is a lot friendlier that the `chip` value. So the following variable query result:
```
```suggestion
For example, when querying the `node_hwmon_chip_names` Prometheus metric, the `chip_name` is a lot friendlier that the `chip` value. So the following variable query result:
```

oddlittlebird commented on docs/sources/variables/filter-variables-with-regex.md:
````diff
@@ -79,4 +79,35 @@ demo.robustperception.io:9090
+enp216s0f0np3         0000:d7:00_0_0000:d8:00_3
+```
+
+**NOTE:** only `text` and `value` capture group names are supported.
````
```suggestion
**Note:** Only `text` and `value` capture group names are supported.
```

oddlittlebird reviewed (commented):
Yes, that is much better! I added a little more polish, but now I can understand what this does. GREAT STUFF!

yenalsnavaj committed:
Update docs/sources/variables/filter-variables-with-regex.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

yenalsnavaj committed:
Apply suggestions from code review

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

oddlittlebird commented:
Thanks, @yenalsnavaj! The code review will be a separate process, done by the developers. We appreciate the documentation coming up front, though!

hugohaggmark commented on public/app/features/variables/query/reducer.ts:
```diff
@@ -86,6 +86,22 @@ const sortVariableValues = (options: any[], sortOrder: VariableSort) => {
 };
 
+const getAllMatches = (str: string, regex: RegExp): any => {
+  let results = {};
+  let matches;
+
+  if (regex.global) {
+    while ((matches = regex.exec(str))) {
+      _.merge(results, matches);
+    }
+  } else {
+    if ((matches = regex.exec(str))) {
+      _.merge(results, matches);
+    }
+  }
+  return results;
```
This looks ok, but having the nested else doesn't seem that clean. What about something like this
```suggestion
    const results = {};
  let matches;

  do {
    matches = regex.exec(str);
    _.merge(results, matches);
  } while (regex.global && matches);

  return results;
```

yenalsnavaj replied:
Agree, much cleaner. Extracting the method was already part of a refactoring. I always need to push a bit further...

hugohaggmark commented on public/app/features/variables/query/reducer.ts:
```diff
@@ -109,13 +125,24 @@ const metricNamesToVariableValues = (variableRegEx: string, sort: VariableSort,
+
+      if (matches.groups && matches.groups.value) {
+        value = matches.groups.value;
+      } else if (matches.groups && matches.groups.hasOwnProperty('value')) {
```
Do I understand this correctly that we ignore any falsy values here? Couldn't the value be assigned the text group value if that exists?

yenalsnavaj replied:
I had considered that -- but as the value of the variable will almost always be used as a key is some other dashboard metric query I felt that it might introduce unwanted behaviour. I'm not set on that though so if you think it would be better we could include. Whatever is most consistent and understandable.

hugohaggmark replied:
I think we already do this in line 117 above, so my thought was to keep that behavior. https://github.com/grafana/grafana/blob/81699f07d3dc3d6706f7ea7e4a2a24a1db06a836/public/app/features/variables/query/reducer.ts#L117

hugohaggmark reviewed (commented):
This is a powerful and great addition @yenalsnavaj 🌟  I've left some things to consider and also I think we need to advertize this powerful new feature somehow.

1. Maybe add more information to the placeholder here:
![image](https://user-images.githubusercontent.com/562238/97954171-12f73e00-1da3-11eb-8afe-91bdd55b3bbc.png)
2. Would be great if icon tooltip here could include a link to the documentation you wrote
![image](https://user-images.githubusercontent.com/562238/97954209-2e624900-1da3-11eb-9dbd-932c503a25c3.png)

yenalsnavaj committed:
Apply suggestions from code review

Co-authored-by: Hugo Häggmark <hugo.haggmark@gmail.com>

yenalsnavaj commented:
> This is a powerful and great addition @yenalsnavaj 🌟 I've left some things to consider and also I think we need to advertize this powerful new feature somehow.
> 
> 1. Maybe add more information to the placeholder here:
>    ![image](https://user-images.githubusercontent.com/562238/97954171-12f73e00-1da3-11eb-8afe-91bdd55b3bbc.png)
> 2. Would be great if icon tooltip here could include a link to the documentation you wrote
>    ![image](https://user-images.githubusercontent.com/562238/97954209-2e624900-1da3-11eb-9dbd-932c503a25c3.png)

I'll take a look at adding something to the tool tip

hugohaggmark commented:
> > This is a powerful and great addition @yenalsnavaj 🌟 I've left some things to consider and also I think we need to advertize this powerful new feature somehow.
> > 
> > 1. Maybe add more information to the placeholder here:
> >    ![image](https://user-images.githubusercontent.com/562238/97954171-12f73e00-1da3-11eb-8afe-91bdd55b3bbc.png)
> > 2. Would be great if icon tooltip here could include a link to the documentation you wrote
> >    ![image](https://user-images.githubusercontent.com/562238/97954209-2e624900-1da3-11eb-9dbd-932c503a25c3.png)
> 
> I'll take a look at adding something to the tool tip

Let me know if something is confusing, I just think this outstanding feature is super-useful, and it's a shame if users can't find it/use it.

yenalsnavaj commented:
> > > This is a powerful and great addition @yenalsnavaj 🌟 I've left some things to consider and also I think we need to advertize this powerful new feature somehow.
> > > 
> > > 1. Maybe add more information to the placeholder here:
> > >    ![image](https://user-images.githubusercontent.com/562238/97954171-12f73e00-1da3-11eb-8afe-91bdd55b3bbc.png)
> > > 2. Would be great if icon tooltip here could include a link to the documentation you wrote
> > >    ![image](https://user-images.githubusercontent.com/562238/97954209-2e624900-1da3-11eb-9dbd-932c503a25c3.png)
> > 
> > 
> > I'll take a look at adding something to the tool tip
> 
> Let me know if something is confusing, I just think this outstanding feature is super-useful, and it's a shame if users can't find it/use it.

Makes sense - I can't see how to include a <Link/> <a> or whatever inside the <InlineFormLabel/> - any pointers?

hugohaggmark commented:
> > > > This is a powerful and great addition @yenalsnavaj 🌟 I've left some things to consider and also I think we need to advertize this powerful new feature somehow.
> > > > 
> > > > 1. Maybe add more information to the placeholder here:
> > > >    ![image](https://user-images.githubusercontent.com/562238/97954171-12f73e00-1da3-11eb-8afe-91bdd55b3bbc.png)
> > > > 2. Would be great if icon tooltip here could include a link to the documentation you wrote
> > > >    ![image](https://user-images.githubusercontent.com/562238/97954209-2e624900-1da3-11eb-9dbd-932c503a25c3.png)
> > > 
> > > 
> > > I'll take a look at adding something to the tool tip
> > 
> > 
> > Let me know if something is confusing, I just think this outstanding feature is super-useful, and it's a shame if users can't find it/use it.
> 
> Makes sense - I can't see how to include a or whatever inside the - any pointers?

I think you can pass an React Element (with text and a link) to the tooltip prop here:
https://github.com/grafana/grafana/blob/45f311874de15e7b54e5e84b4ef2d24755406b2a/public/app/features/variables/query/QueryVariableEditor.tsx#L196

The placeholder can be changed here: https://github.com/grafana/grafana/blob/45f311874de15e7b54e5e84b4ef2d24755406b2a/public/app/features/variables/query/QueryVariableEditor.tsx#L203

yenalsnavaj committed:
Use text capture if value capture does not match

This is to keep the behaviour consistent with the current behavior. See
discussion https://github.com/grafana/grafana/pull/28625/files#r516490942

yenalsnavaj committed:
Improve regex field placeholder and tooltip message

To make the feature more discoverable to users the place holder example
now includes the named capture groups. The tool tip message also
includes a reference and link to the documentation.

yenalsnavaj commented:
@hugohaggmark just noticed that the htlm tags that I had included were stripped out so message didn't make sense. I had located files but wasn't sure how to include a link element - have figured it out now though so please let me know what you think. 

Also includes the updated tests and change in behaviour to use text capture if value capture doesn't match.

hugohaggmark reviewed (approved):
I think this looks great. Thank you for contributing @yenalsnavaj!

AnderssonPeter commented:
Does the 7.3.4 release contain this functionality?
The documentation states that it should exist in the 7.3.X releases https://grafana.com/docs/grafana/latest/variables/filter-variables-with-regex/#filter-and-modify-using-named-text-and-value-capture-groups

But i cant get it to work, i have the following nodes
* `node_meta{container_label_com_docker_swarm_node_id="7ixpxu9rftn1wfexyjfm1whwa",instance="10.0.15.16:9100",job="node-exporter",node_id="7ixpxu9rftn1wfexyjfm1whwa",node_name="owgeotest01.localdomain"}`
* `node_meta{container_label_com_docker_swarm_node_id="b1v7zostjjbl37ly5mronsjju",instance="10.0.15.15:9100",job="node-exporter",node_id="b1v7zostjjbl37ly5mronsjju",node_name="owgeotest02.localdomain"}`

And i have tried with the following regexes and none of them worked
* `/node_id="(?<value>[^"]+)".*node_name="(?<text>[^"]+)"/` - Just shows the id
* `/node_id="(?<value>[^"]+)".*node_name="(?<text>[^"]+)"/g` - just shows the id and i only get 1 result.
* `/node_id="(?<value>[^"]+)"|node_name="(?<text>[^"]+)"/g` - Just shows the id
* `/node_name="(?<text>[^"]+|node_id="(?<value>[^"]+)")"/g` - Just shows one name, value does not work (if i select nothing matches)
* `/node_name="(?<text>[^"]+|node_id="(?<value>[^"]+)")"/` - Shows the name of all nodes but values is lost (if i select nothing matches)

AnderssonPeter commented:
After looking at the commit for QueryVariableEditor.tsx it does not look like this change is included, it would be nice if the documentation had a warning.

hugohaggmark commented:
@AnderssonPeter sorry for the confusion, sometimes our docs get updated before the changes do. Especially when the feature contribution is external, we try to keep everything in the same PR. If you have a look in the milestone, you can see that we schedule this feature for the 7.4.0 realease.
![image](https://user-images.githubusercontent.com/562238/100701600-40360c80-33a0-11eb-8449-c28e51d1ee57.png)

Again sorry for the inconvenience and we'll try to better. You can open a PR that clarifies this in the doc [https://github.com/grafana/grafana/blob/master/docs/sources/variables/filter-variables-with-regex.md](https://github.com/grafana/grafana/blob/master/docs/sources/variables/filter-variables-with-regex.md).

AnderssonPeter commented:
Thanks for clearifying!
